### PR TITLE
cat utility refactoring

### PR DESCRIPTION
**Pull request by kevgeniy**
- Less if branches: declare variables _tab_ and _end_of_line_ in the beginning of the functions and use them instead of **if-s**
- Unnecessary range variable deleted: use simple _i32_ instead
- Optional fail-assertion commented: it is local for cat utility and is always true. At the same time it is in the bottleneck function and increases the run time significantly
- Use rustfmt to structure code

### CODE REVIEW

File: src/cat/cat.rs [hunk 1]
```diff
@@ -296,21 +320,21 @@ impl<'a, W: Write> UnsafeWriter<'a, W> {
     }
 }
 
-#[inline(never)]
-fn fail() -> ! {
-    panic!("assertion failed");
-}
+//#[inline(never)]
```
Comment (mpkh) [inline src/cat/cat.rs]: commented code should be marked with an TODO or message or removed

**Comment by wimh** [comment]
> Optional fail-assertion commented: it is local for cat utility and is always true. At the same time it is in the bottleneck function and increases the run time significantly

While it is true that the assertion will never be called in the current version, it can be if someone changes code elsewhere not knowing this, the software can start to crash, but it can also depend on the input. At least it would be good to add the assertion when compiled in debug mode.

For example if I change `let mut flush_counter = 0usize..1024*32;` because I think it might improve performance, the following command will crash (illegal hardware instruction):

```
dd if=/dev/zero bs=1k count=64  | ./target/release/cat -vn  >/dev/null
```

However these will not:

```
dd if=/dev/zero bs=1k count=64  | ./target/release/cat -v  >/dev/null
dd if=/dev/zero bs=1k count=64  | ./target/release/cat -n  >/dev/null
dd if=/dev/zero bs=1k count=64 | tr '\0' '0'  | ./target/release/cat -vn  >/dev/null
```

So it might not even get noticed. 

But I am also interested which speed improvement you see. 

Using your version I see the following performance (on my pc with just zeros as input):
```
$ dd if=/dev/zero bs=1k count=2M | time ./target/release/cat -vn  >/dev/null
./target/release/cat -vn > /dev/null   11.89s  user 0.38s system 99% cpu 12.270 total

$ dd if=/dev/zero bs=1k count=2M | time ./target/release/cat  >/dev/null
./target/release/cat > /dev/null   0.20s  user 1.63s system 99% cpu 1.840 total

$ dd if=/dev/zero bs=1k count=2M | tr '\0' '0'  | time ./target/release/cat -vn  >/dev/null
./target/release/cat -vn > /dev/null   15.78s  user 0.30s system 99% cpu 16.078 total

$ dd if=/dev/zero bs=1k count=2M | tr '\0' '0'  | time ./target/release/cat   >/dev/null
./target/release/cat > /dev/null   0.23s  user 1.01s system 62% cpu 1.983 total
```

If I add `if len > dst.len()` and related code:

```
        let dst = &mut self.buf[self.pos..];
        let len = buf.len();
        if len > dst.len() {
            fail();
        }
        unsafe { copy_nonoverlapping(buf.as_ptr(), dst.as_mut_ptr(), len) }
```
I see the following result:
```
$ dd if=/dev/zero bs=1k count=2M | time ./target/release/cat -vn  >/dev/null
./target/release/cat -vn > /dev/null   10.85s  user 0.32s system 99% cpu 11.172 total

$ dd if=/dev/zero bs=1k count=2M | time ./target/release/cat  >/dev/null
./target/release/cat > /dev/null   0.21s  user 1.48s system 99% cpu 1.689 total

$ dd if=/dev/zero bs=1k count=2M | tr '\0' '0'  | time ./target/release/cat -vn  >/dev/null
./target/release/cat -vn > /dev/null   12.84s  user 0.26s system 99% cpu 13.099 total

$ dd if=/dev/zero bs=1k count=2M | tr '\0' '0'  | time ./target/release/cat   >/dev/null
./target/release/cat > /dev/null   0.24s  user 0.96s system 59% cpu 2.016 total
```
So just one out of these 4 tests is faster. (but this was just a quick and dirty test).

**Comment by kevgeniy** [comment]
> But I am also interested which speed improvement you see

With `cargo bench` I have always had a 20-40% measurement error. So I used profiler instead and that place (`if` condition) was marked as one of the bottlenecks. I think it is a good reason to remove that code because **coreutils** is supposed to be efficient.

Furthermore, I think that using `panic` to validate the correctness of the written code is ideologically not right for release mode. As I can see, a special _unsafe_ writer was developed and used instead of _standard_ BufWriter solely to increase performance. If so, we _must_ guarantee that it is always used correctly in our code.

But I agree that it is a good idea to use `fail` in debug mode to support bug finding. Fixed this, but I'm quite new to rust and not sure if did it correctly.

**Comment by kevgeniy** [comment]
Rewrote cat to eliminate code duplication and make it **safe**. Run time is changed:
- About the same when -T and -v options are not used
- ~40% less when use -T and do not use -v option
- ~20% more when use -v option

What do you think?

**Comment by wimh** [comment]
I'm not a uutils member, but I would always prefer safe (but also readable) code over fast code. For me that is the promise of rust; always safe, but trying to be as fast as possible. Otherwise I could just use the gnu version (written in C). Note that I didn't worry about the `unsafe` keyword, just about the assertion which was removed. So I think it is great that this can be solved in a safe way.

I noticed a recursive call in `write_tab_to_end`. If I feed 31kb of just tabs, this causes a stack overflow in debug mode:

```
dd if=/dev/zero bs=1k count=31 | tr '\0' '\t'  | ./target/debug/cat -T
```

(In release mode, there is plenty of stack space left). You could consider wrapping the recursive call in [`stacker::maybe_grow`](http://alexcrichton.com/stacker/stacker/index.html) if you want to keep it recursive.

It looks like there are two typos in the comment above `write_to_end`; `symbols` and `bytes - 1`

**Comment by kevgeniy** [comment]
Thank you a lot! Fixed both issues. 

Writing about safety I keep in mind the following:

- `unsafe` code is not used anymore. It is not necessary at all, but is quite nicely.
- UnsafeWriter is replaced by BufWriter, which is well-tested. As one example (bug), consider `drop` method of UnsafeWriter. It uses `flush_buf` method to write the content of the buffer on disk. `flush_buf`, in its turn, uses `write` method instead of `write_all`. This means that current cat version with UnsafeWriter may work incorrectly in certain circumstances.
- BufWriter, in opposite to UnsafeWriter, do not require from programmer to prove the correctness of using`write` and `flush` methods. It makes the code significantly more readable and easy to understand.

File: src/cat/cat.rs [hunk 2]
```diff
@@ -48,280 +45,202 @@ pub fn uumain(args: Vec<String>) -> i32 {
...
+fn write_fast(files: Vec<String>) {
+    let mut writer = stdout();
+    let mut in_buf = [0; 1024 * 64];
 
-        let mut in_buf  = [0; 1024 * 31];
-        let mut out_buf = [0; 1024 * 64];
-        let mut writer = UnsafeWriter::new(&mut out_buf[..], stdout());
-        let mut at_line_start = true;
-        let mut one_blank_kept = false;
+    for (mut reader, _) in files.iter().filter_map(|p| open(&p[..])) {
```
Comment (Arcterus) [inline src/cat/cat.rs]: I'm pretty sure that by opening all the files at once like this you could eventually run out of file handles (I know that it was done like this in the original code).

File: src/cat/cat.rs [hunk 3]
```diff
@@ -48,280 +45,202 @@ pub fn uumain(args: Vec<String>) -> i32 {
...
-    fn flush(&mut self) -> Result<()> {
-        self.flush_buf().and_then(|()| self.inner.flush())
-    }
-}
-
-impl<'a, W: Write> Drop for UnsafeWriter<'a, W> {
-    fn drop(&mut self) {
-        let _ = self.flush_buf();
+        match byte {
+                9 => writer.write_all(tab),
```
Comment (Arcterus) [inline src/cat/cat.rs]: The indentation here is 8 when it should be 4.

File: src/cat/cat.rs [hunk 4]
```diff
@@ -48,280 +45,202 @@ pub fn uumain(args: Vec<String>) -> i32 {
...
+    };
+    // initialize tab simbol
+    let tab = if show_tabs {
+        "^I".as_bytes()
+    } else {
+        "\t".as_bytes()
+    };
     let mut line_counter: usize = 1;
 
     for (mut reader, interactive) in files.iter().filter_map(|p| open(&p[..])) {
```
Comment (Arcterus) [inline src/cat/cat.rs]: Same thing with opening all the files at once here.

File: src/cat/cat.rs (hunk 3 above)
Comment (kevgeniy) [inline src/cat/cat.rs]: Fixed this.

File: src/cat/cat.rs (hunk 2 above)
Comment (kevgeniy) [inline src/cat/cat.rs]: You are not right here for two reasons:

1. `files.iter().filter_map(|p| open(&p[..]))` is no more than `struct` which implement `Map`. On each step of the loop the `next` method is called.
2. `reader` variable is local to loop body, so `reader.drop()` is used at the end of every iteration. As far as I know, `drop` method automatically closes the file.

**Comment by kevgeniy** [comment]
Fixed both issues.

**Comment by kevgeniy** [comment]
Surprisingly, my fork began to fail on one `stat` test. I updated the fork to be up-to-date and now it is different (git diff) from _master_ in only two `cat`'s files. So, I'm not sure what I should (and can) do.

**Comment by wimh** [comment]
See https://travis-ci.org/uutils/coreutils/builds, master fails too (since my pull request was accepted). The last successful build was using rust 1.12.1, the failing builds use rust 1.13.0. The same applies to this fork. So it is obviously unrelated to your changes. Someone has to fix this in a separate pull request. I can look into this, but not until the end of the day. I will create a new issue to track this.

We can see in the travis log why the build is failing, so it should not be a problem. I think you can remove the merge commit.

File: src/cat/cat.rs (hunk 2 above)
Comment (wimh) [inline src/cat/cat.rs]: I agree, I think the following similar test shows this:
```
struct Droppable {
    name: &'static str,
}

impl Drop for Droppable {
    fn drop(&mut self) {
        println!("Dropping {}", self.name);
    }
}

fn main() {
    let a = ["a", "b", "c", "d"];

    for mut i in a.iter().filter_map(|s| Some(Droppable { name: s })) {
        println!("Processing {}", i.name);
    }
}
```
output on https://play.rust-lang.org/:
```
Processing a
Dropping a
Processing b
Dropping b
Processing c
Dropping c
Processing d
Dropping d
```

Comment (Arcterus) [inline src/cat/cat.rs]: Ah, I forgot that `filter_map` made a `struct`.

**Comment by Arcterus** [comment]
I'll merge this after you get rid of the merge commit.

**Pull request merged**